File: hw2/UDP/UDPServer.py
#
# SimpleEchoUDPServer.py
#

from socket import *
import datetime
import time

serverPort = 12000
serverSocket = socket(AF_INET, SOCK_DGRAM)
serverSocket.bind(('', serverPort))
requestNumber = 0
byeMessage = 'Bye bye~'

print("Server is ready to receive on port", serverPort)
runTime = time.time()

while True:
    try:
        message, clientAddress = serverSocket.recvfrom(2048)
        print('Connection requested from', clientAddress)

        message = message.decode()
        requestNumber += 1

        cmd = message.upper()[:11]
        replyMessage = ''

        if cmd == 'ASK_TXTCONV':
            replyMessage = message[12:].upper()
        elif cmd == 'ASK_IP_PORT':
            replyMessage = clientAddress[0] + ',' + str(clientAddress[1])
        elif cmd == 'ASK_REQ_NUM':
            replyMessage = str(requestNumber)
        elif cmd == 'ASK_RUNTIME':
            currentTime = time.time()
            replyMessage = str(datetime.timedelta(seconds=currentTime-runTime)).split(".")[0]
        elif cmd == 'ASK_CONNEND':
            print(byeMessage)
            
        else:
            replyMessage = 'Wrong Command'

        serverSocket.sendto(replyMessage.encode(), clientAddress)

    except KeyboardInterrupt:
        print(byeMessage)
        break
    except ConnectionAbortedError:
        print(byeMessage)
    
    

File: hw2/UDP/server/UDPServer.go
/**
 * UDPServer.go
 **/

package main

import (
	"fmt"
	"net"
	"os"
	"os/signal"
	"strconv"
	"strings"
	"syscall"
	"time"
)

func main() {
	serverPort := "12000"

	pconn, _ := net.ListenPacket("udp", ":"+serverPort)
	fmt.Printf("Server is ready to receive on port %s\n", serverPort)
	start_time := time.Now()
	request_number := 0
	buffer := make([]byte, 1024)

	signals := make(chan os.Signal, 1)
	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		<-signals
		fmt.Println("Bye bye~")
		os.Exit(0)
	}()

	for {
		count, r_addr, _ := pconn.ReadFrom(buffer)
		fmt.Printf("Connection requested from %s %d\n", r_addr, count)
		request_number += 1
		recv_message := string(buffer[:count])
		cmd := strings.ToUpper(recv_message[:11])
		switch cmd {
		case "ASK_TXTCONV":
			reply_message := strings.ToUpper(recv_message[12:])
			fmt.Println(reply_message)
			pconn.WriteTo([]byte(reply_message), r_addr)
		case "ASK_IP_PORT":
			reply_message := r_addr.String()
			fmt.Println(reply_message)
			pconn.WriteTo([]byte(reply_message), r_addr)
		case "ASK_REQ_NUM":
			reply_message := request_number
			pconn.WriteTo([]byte(strconv.Itoa(reply_message)), r_addr)
		case "ASK_RUNTIME":
			reply_message := time.Since(start_time).String()
			pconn.WriteTo([]byte(reply_message), r_addr)
		case "ASK_CONNEND":
			fmt.Println("Bye bye~")
		}
	}
}
